src/bvh/bvh_node.rs

```diff
@@ -0,0 +1,608 @@
+                    / aabb_bounds.surface_area();
+                if cost < min_cost {
+                    min_bucket = i;
+                    min_cost = cost;
+                    child_l_aabb = child_l.aabb;
+                    child_l_centroid = child_l.centroid;
+                    child_r_aabb = child_r.aabb;
+                    child_r_centroid = child_r.centroid;
+                }
+            }
+            // Join together all index buckets.
+            // split input indices, loop over assignments and assign
+            let (l_assignments, r_assignments) = bucket_assignments.split_at_mut(min_bucket + 1);
+
+            let mut l_count = 0;
+            for group in l_assignments.iter() {
+                l_count += group.len();
+            }
+
+            let (child_l_indices, child_r_indices) = indices.split_at_mut(l_count);
+            let mut i = 0;
+            for group in l_assignments.iter() {
+                for x in group {
+                    child_l_indices[i] = *x;
+                    i += 1;
+                }
+            }
+            i = 0;
+            for group in r_assignments.iter() {
+                for x in group {
+                    child_r_indices[i] = *x;
+                    i += 1;
+                }
+            }
+
+            (
+                (child_l_aabb, child_l_centroid, child_l_indices),
+                (child_r_aabb, child_r_centroid, child_r_indices),
+            )
+        })
+    }
+
+    /// Traverses the [`Bvh`] recursively and returns all shapes whose [`Aabb`] is
+    /// intersected by the given [`Ray`].
+    ///
+    /// [`Aabb`]: ../aabb/struct.Aabb.html
+    /// [`Bvh`]: struct.Bvh.html
+    /// [`Ray`]: ../ray/struct.Ray.html
+    ///
+    pub fn traverse_recursive(
+        nodes: &[BvhNode<T, D>],
+        node_index: usize,
+        ray: &Ray<T, D>,
+        indices: &mut Vec<usize>,
+    ) {
+        match nodes[node_index] {
+            BvhNode::Node {
+                ref child_l_aabb,
+                child_l_index,
+                ref child_r_aabb,
+                child_r_index,
+                ..
+            } => {
+                if ray.intersects_aabb(child_l_aabb) {
+                    BvhNode::traverse_recursive(nodes, child_l_index, ray, indices);
+                }
+                if ray.intersects_aabb(child_r_aabb) {
+                    BvhNode::traverse_recursive(nodes, child_r_index, ray, indices);
+                }
+            }
+            BvhNode::Leaf { shape_index, .. } => {
+                indices.push(shape_index);
+            }
+        }
+    }
+}
+
+/// Shapes holds a mutable ptr to the slice of Shapes passed in to build. It is accessed only through a ShapeIndex.
+/// These are a set of unique indices into Shapes that are generated at the start of the build process. Because they
+/// are all unique they guarantee that access into Shapes is safe to do in parallel.
+pub struct Shapes<'a, S> {
+    ptr: *mut S,
```

Comment:
I get a bit bit nervous adding more `unsafe` and raw pointers to our code but I guess it can't be helped in the name of performance. :)

Reply:
Yea I didn't want to use unsafe here either, but the other option was UB because of how we are passing this around.

Reply:
Fair enough.